### PR TITLE
p2p: add a per-message type send and receive metric

### PR DESCRIPTION
**Pull request by williambanfield**
This pull request has one primary goal: to add a new metric for measuring the p2p bytes sent and received by message type.

To do this, parts of the [p2p.Envelope](https://github.com/tendermint/tendermint/blob/b603b4f9f546fa9b8b372ae66072e5c17caae94b/internal/p2p/channel.go#L15) from v0.35.x were resurrected. This was done to change the p2p.Send and associated methods along with the reactor.Receive method from taking a []byte as an argument to taking an object with an unserialized proto.Message as an argument. Having the actual proto.Message in hand in the p2p code allows for the type of the message to be determined. Otherwise, the p2p code cannot know at all what the type of the message being sent is and the only alternative is to litter p2p message size calculations all over the code, which seems much more awkward.

The p2p.Envelope resurrection required a few steps:

proto.Message must be [added to the ChannelDescriptor](https://github.com/tendermint/tendermint/blob/4d78096843bcef3dd408147d2da67fbf99e1ff60/p2p/conn/connection.go#L727) so that the p2p code knows what to unmarshal into when receiving a message.
All Send and Receive methods must be changed to use the Envelope type. This means that the reactors pass a proto message that has not yet been marshaled to the p2p layer and the p2p layer similarly calls the `Receive` method with an unmarshaled proto. All code that marshals and unmarshals the protos in the reactors was removed. Finally, many of the reactors have a wrapper type with a `Sum` field that can contain one of any of the p2p messages that reactor can receive. To actually be able to determine what kind of underlying message was being sent and received for the purposes of the metric, the v0.35 [Wrapper](https://github.com/tendermint/tendermint/blob/b4eaccd242b875511dc758751fe375348300ebbf/internal/p2p/router.go#L115) was resurrected, albeit with a minor change: instead of `Wrap` being implemented on the wrapper type, it's implemented on each of the `Sum` types so that they each know how to Wrap themselves.

With all of these changes in place, the only remaining step was to put the metric into both the send and receive code.

closes: https://github.com/tendermint/tendermint/pull/9599

#### PR checklist

- [ ] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [ ] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

**Review by cmwaters — APPROVED** [review]
Seems all in order

File: p2p/metrics.go [hunk 1]
```diff
@@ -24,4 +36,43 @@ type Metrics struct {
...
+	m.mtx.RUnlock()
+
+	s := t.String()
+	ss := valueToLabelRegexp.FindStringSubmatch(s)
+	l := fmt.Sprintf("%s_%s", ss[1], ss[2])
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	m.messageLabelNames[t] = l
+	return l
+}
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply p2p/metrics.go:71]: Fair enough. This version is pretty generic and can generate a name of any type at the moment. The `Stringer` will generate a string representation of the value when all we really want here is the type name.

File: p2p/peer.go [hunk 2]
```diff
@@ -396,12 +426,29 @@ func createMConnection(
 			// which does onPeerError.
 			panic(fmt.Sprintf("Unknown channel %X", chID))
 		}
+		mt := msgTypeByChID[chID]
+		msg := proto.Clone(mt)
+		err := proto.Unmarshal(msgBytes, msg)
+		if err != nil {
+			panic(fmt.Errorf("unmarshaling message: %s", err))
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply p2p/peer.go]: done.

File: p2p/metrics.go [hunk 3]
```diff
@@ -1,6 +1,11 @@
 package p2p
 
 import (
+	"fmt"
+	"reflect"
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply p2p/metrics.go:5]: Yeah I couldn't find what the reasoning behind blocking this import is

File: p2p/metrics.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmwaters) [reply p2p/metrics.go:71]: I guess you'd need to filter the string anyway to remove special characters. Your implementation seems semantically fine.

**Pull request merged**